runtest.py

```diff
@@ -9,70 +9,97 @@
-    if len(expected) != len(actual):
-        actual_tracebacks: List[str] = list(map(lambda x: str(x['traceback'][0]), actual))
-        print(f"Expected errors: {expected}", file=sys.stderr)
-        print(f"Actual errors: {actual_tracebacks}", file=sys.stderr)
-        return False
-    for e, a in zip(expected, actual):
-        if e not in a['traceback'][0]:
-            print(f"Expected error = {e}, actual error = {a['traceback'][0]}", file=sys.stderr)
-            return False
-    return True
+
+def check_errors(file_name, expected: List[str], actual: List[Any]) -> bool:
+    """When errors occur they are due to a mismatch in the errors that occurred at runtime
+    and the expected error that are defined in notebooks
+    This produces huge output but the relevant information is between bars of '=' at the end.
+    Look at what was expected and either fix the error or add the text of the error that occurred
+    :note Errors shown may have hidden escape sequence for colors etc. When pasting be careful about this.
+
+    :param file_name: name of file where errors occurred
+    :param expected:  errors that were expected
+    :param actual:    errors that occurred.
+    :return:
+    """
+    actual_tracebacks: List[str] = list(map(lambda x: str(x['traceback'][0][:100]), actual))
+
+    return_value = True
+    for i, (e, a) in enumerate(itertools.zip_longest(expected, actual_tracebacks, fillvalue="-- No Error --")):
+        if e not in a:
+            if return_value:
+                print("=" * 100, file=sys.stderr)
+                print(f"Errors detected in {file_name}", file=sys.stderr)
+            print(f"No match for {i}-th expected error '{e}' got '{a[:100]}'", file=sys.stderr)
+            return_value = False
+    if not return_value:
+        print("=" * 100, file=sys.stderr)
+    return return_value
+
 
 notebooks: Dict[str, List[str]] = {
-    # For some reason the NotImplementedError gets swallowed up by
-    # setupFirrtlTerpBackend.
-    # Instead we have to try and catch the "java.lang.Exception: Problem with compilation"
-    # message instead.
+    # This is the list of pages to try and rum along with each pages expected errors.
 
     "1_intro_to_scala.ipynb": [],
-    "2.1_first_module.ipynb": ['scala.NotImplementedError'],
-    "2.2_comb_logic.ipynb": ['Compilation Failed'] + ['java.lang.Exception: Problem with compilation'] * 3,
-    "2.3_control_flow.ipynb": ['scala.NotImplementedError'] * 2 + ['Compilation Failed']
-      + ['scala.NotImplementedError'] + ['Compilation Failed'],
-    "2.4_sequential_logic.ipynb": ['java.lang.Exception: Problem with compilation'] * 2,
-    "2.5_exercise.ipynb": ['java.lang.Exception: Problem with compilation'] * 4,
-    "2.6_testers2.ipynb": [],
+    "2.1_first_module.ipynb": [],
+    "2.2_comb_logic.ipynb": ['Compilation Failed'] +
+                            ['chisel3.internal.ChiselException: Exception thrown when elaborating ChiselGeneratorAnnotation'] +
```

Comment:
It's unfortunate there's not a way to grab the underlying error.

Reply:
Yes, there's a lot of useful work that could be done here. I feel like a made a lot of useful changes in interpreting the output. But more work would be good. It would be cool if we could mark cells that required work and not run them